docs/docs/python/integrations/google-cloud-functions.mdx

```diff
@@ -67,6 +66,42 @@ Use [layout](../../../reference/targets/python_google_cloud_function.mdx#layout)
 `file` / `files` targets will not be included in the built Cloud Function because filesystem APIs like `open()` would not load them as expected. Instead, use the `resource` / `resources` target. See [Assets and archives](../../using-pants/assets-and-archives.mdx) for further explanation.
 :::
 
+### Specifying a runtime explicitly
+
+Pants will attempt to infer the appropriate runtime and its platform-specific details from [the relevant interpreter constraints](../overview/interpreter-compatibility.mdx). For example, given `interpreter_constaints = ["==3.12.*"]`, Pants will build the Lambda package to appropriate for the  `python312` runtime. The Lambda environment is a relatively stable, running Linux on a x86-64 CPU, and thus Pants includes default "complete platforms" specifications that allow choosing appropriate artifacts of third-party dependencies with native code. These are chosen based on the (inferred or explicit) runtime.
+
+The inference may not work in your scenario and/or may not be appropriate:
+
+1. if the interpreter constraints are not obviously referring to all versions from a given minor release series, Pants won't infer the runtime. For instance, `>=3.11,<3.13` is too wide (it covers both 3.11 and 3.12), while `==3.12.1` is too specific (GCF's `python312` runtime may not use that patch version).
+2. if the interpreter constraints refer to a runtime that Pants doesn't have a default complete platforms JSON specification for, Pants will require you to specify one.
+3. if the cloud environment has changed, Pants' default complete platforms value may imply correct artifacts dependency and you may want to customise.
+
+In the first case, [the `runtime` field](../../../reference/targets/python_google_cloud_function.mdx#runtime) can be used to specify which runtime is intended for each `python_google_cloud_function` target. It should be one of the values from [https://cloud.google.com/functions/docs/concepts/python-runtime](https://cloud.google.com/functions/docs/concepts/python-runtime).
+
+```python title="BUILD"
+python_google_cloud_function(
+    name="cloud_function",
+    handler="google_cloud_function_example.py:example_handler",
+    type="event",
+    # Explicit runtime:
+    runtime="python312",
+)
+```
```

Comment:
I guess the sort of "intended" workflow implied here is that the user can rely on interpreter constraints first, falling back to runtime and then complete platforms. I've never found this to be effective 😅 

I pretty much always specify the runtime or complete platforms explicitly because my interpreter constraints support multiple versions of Python.

Reply:
Ah I see. It's always been effective in my experience, where we do just use one version of Python... but I'll think about how to tweak it.

Reply:
IMO, prompting the user to explicitly specify the runtime is a good thing and avoids ambiguity, but that's just my 2c - I think it's okay to fall back on interpreter constraints but that should be a backup option.

I guess the main way this could blow up is if the interpreter constraints and the explicit runtime don't match. Is there another consideration that you had in mind?

Reply:
> IMO, prompting the user to explicitly specify the runtime is a good thing and avoids ambiguity, but that's just my 2c - I think it's okay to fall back on interpreter constraints but that should be a backup option.

In my mind, there's two bits of value in the inference:

- avoiding redundancy (i.e. single point of truth)
- making initial onboarding easy

> I guess the main way this could blow up is if the interpreter constraints and the explicit runtime don't match.

Huh, I don't know what this does in practice 🤔 

> Is there another consideration that you had in mind?

Nope!

---

Also, sorry, I was probably ambiguous in my initial response, I definitely agree that we shouldn't emphasise the inference as much as this first version did (iterating now), just observing that it's been working fine for my real work repos 😄